### PR TITLE
Allow for `null` and `empty` parameters in the MultiField annotation.

### LINKED ISSUE #2952: Support storeNullValue and storeEmptyValue for property that is annotated with @MultiField annotation
Currently storeNullValue and storeEmptyValue are supported for property that is annotated with the @Field annotation. Need to also add support for property that is annotated with @MultiField annotation. Suggesting the following code change in the SimpleElasticsearchPersistentProperty class:

In the SimpleElasticsearchPersistentProperty constructer, update the assignment of storeNullValue and storeEmptyValue to also check the mainField parameters in the @MultiField annotation:

`storeNullValue = (isField && getRequiredAnnotation(Field.class).storeNullValue()) || (isAnnotationPresent(MultiField.class) && getRequiredAnnotation(MultiField.class).mainField().storeNullValue());
storeEmptyValue = isField ? getRequiredAnnotation(Field.class).storeEmptyValue() : isAnnotationPresent(MultiField.class) ? getRequiredAnnotation(MultiField.class).mainField().storeEmptyValue() : true;`

### PR DESCRIPTION
**Pull request by youssef3wi**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] **There is a ticket in the bug tracker for the project in our [issue tracker](https://github.com/spring-projects/spring-data-elasticsearch/issues)**. Add the issue number to the _Closes #issue-number_ line below
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [ ] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Closes #2952

### CODE REVIEW

**Comment by youssef3wi** [comment]
I think it would be better to include these parameters in https://github.com/spring-projects/spring-data-elasticsearch/blob/d079a59cb4568a206764f522bb3fd4ca44af59c5/src/main/java/org/springframework/data/elasticsearch/annotations/InnerField.java#L181
```java
boolean storeNullValue() default false;
boolean storeEmptyValue() default true;
``` 

 to also support inner fields

https://github.com/spring-projects/spring-data-elasticsearch/blob/d079a59cb4568a206764f522bb3fd4ca44af59c5/src/main/java/org/springframework/data/elasticsearch/core/mapping/SimpleElasticsearchPersistentProperty.java#L113

```java
storeNullValue = isField ? getRequiredAnnotation(Field.class).storeNullValue() 
: isMultiField && (getRequiredAnnotation(MultiField.class).mainField().storeNullValue() 
|| Arrays.stream(getRequiredAnnotation(MultiField.class).otherFields())
.anyMatch(InnerField::storeNullValue));

storeEmptyValue = isField ? getRequiredAnnotation(Field.class).storeEmptyValue() 
: !isMultiField || getRequiredAnnotation(MultiField.class).mainField().storeEmptyValue() 
|| Arrays.stream(getRequiredAnnotation(MultiField.class).otherFields())
.anyMatch(InnerField::storeEmptyValue);
```

**Review by sothawo — CHANGES_REQUESTED** [review]
Thanks for taking that over.

As storeEmpty and nul values are written to the index mapping I think there should be a test in the `(Reacive)MappingBuilderUnitTests` as well.

**Review by sothawo — CHANGES_REQUESTED** [review]
The inner field of a multi field can have a different null_value than the outer field. This should be reflected in the mapping builder test, I added the changes for that; it is already working but wasn't in the test.

I think this should be covered in the integration tests as well, using different null values for the outer and inner fields.

And no, I don't know why one would set different values, but it is possible so we need to make sure to support it.

From https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-fields.html:
>  A multi-field doesn’t inherit any mapping options from its parent field.

File: src/test/java/org/springframework/data/elasticsearch/core/index/MappingBuilderUnitTests.java [hunk 1]
```diff
@@ -2570,5 +2597,11 @@ private static class MultiFieldMappedNameEntity {
 		@MultiField(mainField = @Field(type = FieldType.Text, mappedTypeName = "match_only_text"), otherFields = { @InnerField(suffix = "lower_case",
 				type = FieldType.Keyword, normalizer = "lower_case_normalizer", mappedTypeName = "constant_keyword") }) private String description;
 	}
+
+	private static class MultiFieldWithNullEmptyParameters {
+		@Nullable
+		@MultiField(mainField = @Field(name = "empty-field", type = FieldType.Keyword, nullValue = "EMPTY",
+				storeNullValue = true, storeEmptyValue = false)) private List<String> emptyField;
+	}
```
[the remark this answers is not in the document]
Comment (youssef3wi) [reply src/test/java/org/springframework/data/elasticsearch/core/index/MappingBuilderUnitTests.java:2605]: `null_value` only for keyword.

https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#text-params

**Comment by sothawo** [comment]
I just tried to use this in a sample application, but it fails when trying to create the index with the `null_value` mapping in https://github.com/spring-projects/spring-data-elasticsearch/blob/main/src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java#L258. 

Might be an error on our side, or an error in the Elasticsearch client. Won't have any more time today to check.

We should have an integration test for this that creates an index with the mapping, stores a value and retrieves it.

**Comment by sothawo** [comment]
TIL: you cannot set a `null_value` on a field of type `text`. This fails on Elasticsearch level, not only in the Elasticsearch client library.

Although the [documentation about null_value](https://www.elastic.co/guide/en/elasticsearch/reference/current/null-value.html) does not mention this.

Will get back to your PR tomorrow.

**Pull request merged**